File: pixie/eval/agent_evaluator.py
"""Factory for agent-deferred evaluators.

An agent evaluator is a concrete ``Evaluator`` implementation whose
``__call__`` raises :class:`AgentEvaluationPending` instead of returning
an :class:`~pixie.eval.evaluation.Evaluation`.  The test runner catches
this exception and records a :class:`~pixie.harness.run_result.PendingEvaluation`
in the entry result.

Usage::

    from pixie import create_agent_evaluator

    response_quality = create_agent_evaluator(
        name="ResponseQuality",
        criteria=(
            "The response directly addresses the user's question with "
            "accurate, well-structured information."
        ),
    )
"""

from __future__ import annotations

from pixie.eval.evaluable import Evaluable
from pixie.eval.evaluation import Evaluation


class AgentEvaluationPending(Exception):
    """Raised by agent evaluators to signal deferred grading.

    The test runner catches this and records a
    :class:`~pixie.harness.run_result.PendingEvaluation`.

    Attributes:
        evaluator_name: Display name for the evaluator.
        criteria: Grading instructions for the agent.
    """

    def __init__(self, evaluator_name: str, criteria: str) -> None:
        self.evaluator_name = evaluator_name
        self.criteria = criteria
        super().__init__(f"Agent evaluation pending: {evaluator_name}")


class _AgentEvaluator:
    """Evaluator that defers grading to a coding agent."""

    def __init__(self, name: str, criteria: str) -> None:
        self._name = name
        self._criteria = criteria

    @property
    def name(self) -> str:
        """Return the evaluator's display name."""
        return self._name

    async def __call__(self, evaluable: Evaluable) -> Evaluation:
        """Always raises :class:`AgentEvaluationPending`."""
        raise AgentEvaluationPending(self._name, self._criteria)


def create_agent_evaluator(
    name: str,
    criteria: str,
) -> _AgentEvaluator:
    """Create an evaluator whose grading is deferred to a coding agent.

    The returned evaluator satisfies the ``Evaluator`` protocol but
    always raises :class:`AgentEvaluationPending` when called.  The
    test runner catches this and records a pending evaluation.

    Args:
        name: Display name (shown in the scorecard).
        criteria: What to evaluate — the grading instructions the agent
            will follow when reviewing test results and traces.

    Returns:
        An evaluator callable that raises ``AgentEvaluationPending``.
    """
    return _AgentEvaluator(name=name, criteria=criteria)

File: pixie-ts/src/eval/agentEvaluator.ts
/**
 * Factory for agent-deferred evaluators.
 *
 * An agent evaluator raises AgentEvaluationPending instead of returning
 * an Evaluation. The evaluation harness catches this and records a
 * PendingEvaluation in the entry result.
 */

import type { Evaluable } from "./evaluable.js";
import type { Evaluation } from "./evaluation.js";

/**
 * Error raised by agent evaluators to signal deferred grading.
 */
export class AgentEvaluationPending extends Error {
  public readonly evaluatorName: string;
  public readonly criteria: string;

  constructor(evaluatorName: string, criteria: string) {
    super(`Agent evaluation pending: ${evaluatorName}`);
    this.name = "AgentEvaluationPending";
    this.evaluatorName = evaluatorName;
    this.criteria = criteria;
  }
}

class AgentEvaluator {
  private readonly _name: string;
  private readonly _criteria: string;

  constructor(name: string, criteria: string) {
    this._name = name;
    this._criteria = criteria;
  }

  get name(): string {
    return this._name;
  }

  async __call__(evaluable: Evaluable): Promise<Evaluation> {
    void evaluable;
    throw new AgentEvaluationPending(this._name, this._criteria);
  }

  /** Make the instance callable. */
  get call(): (evaluable: Evaluable) => Promise<Evaluation> {
    return this.__call__.bind(this);
  }
}

/**
 * Create an evaluator whose grading is deferred to a coding agent.
 *
 * The returned evaluator satisfies the Evaluator interface but always
 * throws AgentEvaluationPending when called.
 */
export function createAgentEvaluator(
  name: string,
  criteria: string,
): (evaluable: Evaluable) => Promise<Evaluation> {
  const evaluator = new AgentEvaluator(name, criteria);
  const fn = evaluator.call;
  Object.defineProperty(fn, "name", { value: name });
  return fn;
}
